core/src/main/java/org/apache/gravitino/listener/PolicyEventDispatcher.java

```diff
@@ -0,0 +1,374 @@
+import org.apache.gravitino.listener.api.event.DisablePolicyEvent;
+import org.apache.gravitino.listener.api.event.DisablePolicyFailureEvent;
+import org.apache.gravitino.listener.api.event.DisablePolicyPreEvent;
+import org.apache.gravitino.listener.api.event.EnablePolicyEvent;
+import org.apache.gravitino.listener.api.event.EnablePolicyFailureEvent;
+import org.apache.gravitino.listener.api.event.EnablePolicyPreEvent;
+import org.apache.gravitino.listener.api.event.GetPolicyEvent;
+import org.apache.gravitino.listener.api.event.GetPolicyFailureEvent;
+import org.apache.gravitino.listener.api.event.GetPolicyForMetadataObjectEvent;
+import org.apache.gravitino.listener.api.event.GetPolicyForMetadataObjectFailureEvent;
+import org.apache.gravitino.listener.api.event.GetPolicyForMetadataObjectPreEvent;
+import org.apache.gravitino.listener.api.event.GetPolicyPreEvent;
+import org.apache.gravitino.listener.api.event.ListMetadataObjectsForPolicyEvent;
+import org.apache.gravitino.listener.api.event.ListMetadataObjectsForPolicyFailureEvent;
+import org.apache.gravitino.listener.api.event.ListMetadataObjectsForPolicyPreEvent;
+import org.apache.gravitino.listener.api.event.ListPoliciesEvent;
+import org.apache.gravitino.listener.api.event.ListPoliciesFailureEvent;
+import org.apache.gravitino.listener.api.event.ListPoliciesPreEvent;
+import org.apache.gravitino.listener.api.event.ListPolicyInfosEvent;
+import org.apache.gravitino.listener.api.event.ListPolicyInfosFailureEvent;
+import org.apache.gravitino.listener.api.event.ListPolicyInfosForMetadataObjectEvent;
+import org.apache.gravitino.listener.api.event.ListPolicyInfosForMetadataObjectFailureEvent;
+import org.apache.gravitino.listener.api.event.ListPolicyInfosForMetadataObjectPreEvent;
+import org.apache.gravitino.listener.api.event.ListPolicyInfosPreEvent;
+import org.apache.gravitino.listener.api.info.PolicyInfo;
+import org.apache.gravitino.meta.PolicyEntity;
+import org.apache.gravitino.policy.Policy;
+import org.apache.gravitino.policy.PolicyChange;
+import org.apache.gravitino.policy.PolicyContent;
+import org.apache.gravitino.policy.PolicyDispatcher;
+import org.apache.gravitino.utils.NameIdentifierUtil;
+import org.apache.gravitino.utils.PrincipalUtils;
+
+/**
+ * {@code PolicyEventDispatcher} is a decorator for {@link PolicyDispatcher} that not only delegates
+ * policy operations to the underlying policy dispatcher but also dispatches corresponding events to
+ * an {@link EventBus} after each operation is completed. This allows for event-driven workflows or
+ * monitoring of policy operations.
+ */
+public class PolicyEventDispatcher implements PolicyDispatcher {
+  private final EventBus eventBus;
+  private final PolicyDispatcher dispatcher;
+
+  public PolicyEventDispatcher(EventBus eventBus, PolicyDispatcher dispatcher) {
+    this.eventBus = eventBus;
+    this.dispatcher = dispatcher;
+  }
+
+  private PolicyInfo toPolicyInfo(PolicyEntity policy) {
```

Comment:
Can you move this private method to after the public methods? This will make review easier.